src/opt/ReplaceHalfMaskedMemOps.cpp

```diff
@@ -0,0 +1,216 @@
+//      <8 x i1> <i1 true, i1 true, i1 true, i1 true, i1 false, i1 false, i1 false, i1 false>,
+//      <8 x float> <float poison, float poison, float poison, float poison, float 0, float 0, float 0, float 0>)
+//
+// with
+//
+// %ptr = bitcast float* %A to <4 x float>*
+// %vec.half = load <4 x float>, <4 x float>* %ptr
+// %vec = shufflevector <4 x float> %vec.half, <4 x float> <float 0, float 0, float 0, float 0>,
+//                      <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
+//
+void lReplaceMaskedLoad(llvm::IRBuilder<> &B, llvm::CallInst *CI) {
+    llvm::Function *F = CI->getCalledFunction();
+    llvm::Value *ptr = CI->getOperand(0);
+    llvm::Value *passthrough = CI->getOperand(3);
+
+    B.SetInsertPoint(CI);
+
+    passthrough = lShrinkConstVec(F->getParent()->getContext(), passthrough);
+    ptr = lBitcastPointerType(B, ptr, passthrough);
+
+    auto *vecType = passthrough->getType();
+    llvm::Twine origName = CI->getName();
+    llvm::Value *half = B.CreateLoad(vecType, ptr, origName + ".half");
+
+    llvm::Value *replacement = lMergeVectors(B, half, passthrough, origName);
+
+    LLVMCopyMetadata(replacement, CI);
+    CI->replaceAllUsesWith(replacement);
+    CI->eraseFromParent();
+}
+
+llvm::PreservedAnalyses ReplaceHalfMaskedMemOpsPass::run(llvm::Function &F, llvm::FunctionAnalysisManager &FAM) {
+    if (F.empty()) {
+        return llvm::PreservedAnalyses::all();
+    }
+
+    llvm::IRBuilder<> builder(F.getParent()->getContext());
+    std::vector<llvm::CallInst *> storesToReplace;
+    std::vector<llvm::CallInst *> loadsToReplace;
+    for (auto &BB : F) {
+        for (auto &I : BB) {
+            auto *CI = llvm::dyn_cast<llvm::CallInst>(&I);
+            if (!CI) {
+                continue;
+            }
+
+            llvm::Function *CF = CI->getCalledFunction();
+            if (!(CF && CF->isIntrinsic())) {
+                continue;
+            }
+
+            if (CF->getIntrinsicID() == llvm::Intrinsic::masked_store) {
+                llvm::Value *mask = CI->getOperand(3);
+                if (lIsSecondHalfAllFalse(mask)) {
+                    // TODO! sane debug output
+                    // TODO! remarks
+                    // printf("SHRINK masked.store %s\n", f->getName().str().c_str());
+                    storesToReplace.push_back(CI);
+                }
+            }
+
+            if (CF->getIntrinsicID() == llvm::Intrinsic::masked_load) {
+                llvm::Value *mask = CI->getOperand(2);
+                if (lIsSecondHalfAllFalse(mask)) {
+                    // printf("SHRINK masked.load %s\n", f->getName().str().c_str());
+                    loadsToReplace.push_back(CI);
+                }
+            }
+        }
+    }
+
+    for (auto CI : storesToReplace) {
+        lReplaceMaskedStore(builder, CI);
+    }
+
+    for (auto CI : loadsToReplace) {
+        lReplaceMaskedLoad(builder, CI);
+    }
```

Comment:
Should `llvm::PreservedAnalyses::all()` be returned in case both `storesToReplace` and `loadsToReplace` are empty?

Reply:
Done